### ISSUE TITLE
ICU custom translation ids

### ISSUE THREAD

**Issue by LukasAV** [angular/angular#48668]
### Which @angular/* package(s) are relevant/related to the feature request?

localize

### Description

I would like to use custom IDs on ICU expression translations.

The following template

`<h1 i18n="@@IcuPluralTest">ICU plural test {test, plural, =0 {now} other {{{test}} minutes}}</h1>`

Results in two translations in json output format:

```
"IcuPluralTest": "ICU plural test {$ICU}",
"5042638883563958958": "{VAR_PLURAL, plural, =0 {now} other {{INTERPOLATION} minutes}}",
```

However, I have no way to assign a custom ID to the nested ICU expression

### Proposed solution

Enable a syntax like

`<h1 i18n="@@IcuPluralTest">ICU plural test {:@@IcuPluralTestNested:test, plural, =0 {now} other {{{test}} minutes}}</h1>`

that allows assigning an ID to the nested ICU expression

### Alternatives considered

As a workaround, a simple interpolation can be used in the template

`<h1 i18n="@@IcuPluralTest">ICU plural test {{testPlural}}</h1>`

with the ICU expression being defined in the .ts file

```
testPlural = $localize`:@@IcuPluralTestNested:{${this.test}, plural, =0 {now} other {${this.test} minutes}}`;
```

**Comment by schmkr** [comment]
Having the ICU translation parts extracted with random ids makes them loose context. The solution suggested seems nice, but I can also imagine it will quite quickly lead to quite complex translation strings in the template (or ts-files when used with $localize).

An alternative suggestion though, maybe the extracted ICU strings could be prefixed with the translation id they are part of? E.g. in case of the above example:

```
"IcuPluralTest": "ICU plural test {$ICU}",
"IcuPluralTest_5042638883563958958": "{VAR_PLURAL, plural, =0 {now} other {{INTERPOLATION} minutes}}",
```

**Comment by michaelwildvarian** [comment]
This is quite annoying as it keeps trashing my `messages*.xlf` files and makes life unnecessarily hard for translators. The fact alone that the ICU expression is extracted to a separate translation unit makes it challenging. But then having a translation unit with a seemingly random ID that is not linked to the original translation unit in any transparent way makes things plain confusing.

**Comment by NogaMan** [comment]
Why ICU expressions are extracted as separate messages in the first place? This is improper to assume that ICU pieces don't depend on the context of their usage. There are lots of examples in many languages when two seemingly similar ICUs will have different declinations/articles depending on their role in the sentence.
If the Angular team provided a way to disable such extraction into separate messages, they would fix 2-3 related issues at once. Not to say that is the only true way to extract ICUs. They should be extracted together with the context and be unique to each place they are used and not be merged with every other similar ICUs.
Do you have some plans to get rid of this behaviour?

**Comment by th0r** [comment]
> Why ICU expressions are extracted as separate messages in the first place? This is improper to assume that ICU pieces don't depend on the context of their usage. There are lots of examples in many languages when two seemingly similar ICUs will have different declinations/articles depending on their role in the sentence.

I completely agree. Maybe someone from ng team could clarify the reason behind this decision?

**Comment by NogaMan** [comment]
This bug can potentially be fixed along with https://github.com/angular/angular/issues/65143
